include/boost/url/impl/segments_encoded.ipp

```diff
@@ -18,25 +18,158 @@
 namespace boost {
 namespace urls {
 
-string_view
 segments_encoded::
 iterator::
-operator*() const noexcept
+iterator(
+    string_view s,
+    std::size_t nseg) noexcept
+    : begin_(s.data())
+    , pos_(s.data())
+    , next_(s.data())
+    , end_(s.data() + s.size())
 {
-    BOOST_ASSERT(i_ < u_->nseg_);
-    auto p0 = u_->segment(i_);
-    auto const p1 =
-        u_->segment(i_ + 1);
-    if(i_ > 0)
-        ++p0;
+    if(nseg == 0)
+    {
+        next_ = nullptr;
+        return;
+    }
+    auto const n =
+        detail::path_prefix(s);
+    begin_ += n;
+    next_ += n;
+    pos_ += n;
+    auto const i = string_view(
+        begin_, s.size() - n
+            ).find_first_of('/');
+    if(i != string_view::npos)
+        next_ += i;
     else
-        p0 += detail::path_prefix(
-            u_->get(id_path));
-    string_view s(
-        u_->cs_ + p0, p1 - p0);
-    return s;
+        next_ = end_;
+    s_ = string_view(
+        pos_, next_ - pos_);
 }
 
+segments_encoded::
+iterator::
+iterator(
+    string_view s,
+    std::size_t nseg,
+    int) noexcept
+    : i_(nseg)
+    , begin_(s.data())
+    , pos_(s.data() + s.size())
+    , end_(s.data() + s.size())
+{
+    auto const n =
+        detail::path_prefix(s);
+    begin_ += n;
+}
+
+
+auto
+segments_encoded::
+    iterator::
+    operator++() noexcept ->
+    iterator&
+{
+    using bnf_t =
+        path_rootless_rule;
+    BOOST_ASSERT(next_ != nullptr);
+    ++i_;
+    pos_ = next_;
+    error_code ec;
+    // "/" segment
+    pct_encoded_str t;
+    bnf_t::increment(
```

Comment:
We can do better than this though. Since the string was already validated, we can skip the checking of percent-encoded characters and just do a find for the next slash.

Reply:
Yep. I thought about that but didn't want to mess with your logic.

Reply:
I experimented with `find_first_of('/')`. The problem was `find_first_of('/')` plus the edge cases (`next_ == end_`, etc...) became very similar to just replicating the logic from `bnf_t::increment`.